File: lecture09/exercise09-D/solution.cpp
// Exercise 09-D: Invert Binary Tree

#include <iostream>
#include <memory>
#include <queue>
#include <string>
#include <sstream>
#include <vector>

using namespace std;

// Node structures

template <typename T>
struct Node {
    T value;
    struct Node *left;
    struct Node *right;
};

typedef Node<int> IntNode;

// Functions

bool read_values(vector<int> &values) {
    values.clear();

    string line;
    if (getline(cin, line)) {
    	stringstream ss(line);
    	int i;

    	while (ss >> i) {
    	    values.push_back(i);
	}
    }

    return !values.empty();
}

IntNode*  read_tree(vector<int> &values, size_t index=0) {
    if (index >= values.size()) {
    	return nullptr;
    }

    return new Node<int> {
    	values[index],
    	read_tree(values, 2*index + 1),
    	read_tree(values, 2*index + 2)
    };
}

void	dump_tree(IntNode *root) {
    queue<IntNode *> nodes;
    bool first = true;

    nodes.push(root);

    while (!nodes.empty()) {
    	auto node = nodes.front(); nodes.pop();

    	if (!node) {
    	    continue;
	}

    	if (!first) {
    	    cout << ",";
	} else {
    	    first = false;
	}
    	cout << node->value;

    	nodes.push(node->left);
    	nodes.push(node->right);
    }

    cout << endl;
}

IntNode*    invert_tree(IntNode *root) {
    // Base Case: Invalid Node
    if (!root) {
    	return nullptr;
    }

    // Divide and Conquer: Swap left and right and recurse
    swap(root->left, root->right);

    root->left  = invert_tree(root->left);
    root->right = invert_tree(root->right);

    return root;
}

// Main execution

int main(int argc, char *argv[]) {
    vector<int> values;

    while (read_values(values)) {
    	auto root = read_tree(values);
    	invert_tree(root);
    	dump_tree(root);
    }

    return 0;
}

File: lecture09/exercise09-D/solution.py
#!/usr/bin/env python3

# Exercise 09-D: Invert Binary Tree

from collections import deque
from dataclasses import dataclass
from typing      import Optional
import sys

# Classes

@dataclass
class Node:
    value:  int
    left:   Optional['Node']
    right:  Optional['Node']

# Functions

def read_tree(values: list[int], index: int=0) -> Optional[Node]:
    if index >= len(values):
        return None

    return Node(
        values[index], 
        read_tree(values, 2*index + 1), 
        read_tree(values, 2*index + 2)
    )

def walk_tree(root: Optional[Node]) -> list[int]:
    # BFS Traversal with Queue
    nodes  = deque([root])
    values = []

    while nodes:
        node = nodes.popleft()

        if not node:
            continue

        values.append(node.value)

        nodes.append(node.left)
        nodes.append(node.right)

    return values

def dump_tree(root: Optional[Node]) -> None:
    print(','.join(map(str, walk_tree(root))))

def invert_tree(root: Optional[Node]) -> Optional[Node]:
    # Base Case: Invalid Node
    if root is None:
        return None

    # Divide and Conquer: Swap left and right and recurse
    root.left, root.right = invert_tree(root.right), invert_tree(root.left)
    return root

# Main Execution

def main() -> None:
    for line in sys.stdin:
        values = list(map(int, line.split()))
        root   = read_tree(values)
        #dump_tree(root)

        invert_tree(root)
        dump_tree(root)

if __name__ == '__main__':
    main()
